### ISSUE TITLE
Is it possible to navigate from outside of a react component?

### ISSUE THREAD

**Issue by digitalmaster** [react-navigation/react-navigation#1439]
Haven't been able to figure out a way to navigate from outside react-component.

My use case is that I have a saga that waits for a specific redux action to be dispatched; when triggered I want to navigate to a specific route right from my saga function. 

Is this even possible without having to integrate redux and take over state management (trying to keep things simple)?

**Comment by keesee** [comment]
I'm in the process of evaluating react-navigation and am wondering the same.    The app we are starting soon needs several nested navigators so I'd rather not need a bunch of reducers. just for managing the router.  The redux solutions I've skimmed on in react-navigation issues seem a bit hacky.   A router should just work, and you should be able to dispatch an action from anywhere.   I did a quick redux setup as a test and it worked well until I introduced a nested router.   Documentation didn't help.

**Comment by digitalmaster** [comment]
@keesee Yup.. i'm 100% aligned with your description of the complexity i'm trying to avoid. Ideally there should be a way to dispatch actions directly to the react-navigation internal store; maybe by exposing its internal `dispatch` fn so that we just trigger navigationActions directly to it? Or is there a reducer example that works (including nested routes)?

**Comment by Moebats** [comment]
I would like to know this too.  I am trying to navigate to a different scene from within an action dispatch in an actions file.  I am using  dispatch(NavigationActions.navigate({ routeName: 'route' }));
` with the NavigationActions imported up top.  It does not seem to be working though.

**Comment by daose** [comment]
possible duplicate of https://github.com/react-community/react-navigation/issues/840, the solution I mentioned in that issue does require redux integration ([react-navigation redux integration guide](https://reactnavigation.org/docs/guides/redux))

**note:
>Keep in mind that when a navigator is given a navigation prop, it relinquishes control of its internal state. That means you are now responsible for persisting its state, handling any deep linking, integrating the back button, etc.

**Comment by digitalmaster** [comment]
Not exactly duplicate... my question specifically asks for a solution not requiring redux integration. Like an external api to trigger navigation actions which affect its internal store.

So i'm assuming the answer is No.

**Comment by ericvicenti** [comment]
It should be possible to call `.dispatch(action)` on your top-level navigator, provided you aren't providing your own navigation prop as you would with a redux integration

**Comment by unexge** [comment]
you can pass your top-level navigator ref to a service, and dispatch actions from that service.

here is mine:

```js
// App.js

import NavigatorService from './services/navigator';

const Navigator = StackNavigator({ /* ... */ })

class App extends Component {
  // ...

  render(): {
    return (
      <Navigator
        ref={navigatorRef => {
          NavigatorService.setContainer(navigatorRef);
        }}
      />
    );
  }
}
```

```js
// services/navigator.js
// @flow

import { NavigationActions } from 'react-navigation';
import type { NavigationParams, NavigationRoute } from 'react-navigation';

let _container; // eslint-disable-line

function setContainer(container: Object) {
  _container = container;
}

function reset(routeName: string, params?: NavigationParams) {
  _container.dispatch(
    NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({
          type: 'Navigation/NAVIGATE',
          routeName,
          params,
        }),
      ],
    }),
  );
}

function navigate(routeName: string, params?: NavigationParams) {
  _container.dispatch(
    NavigationActions.navigate({
      type: 'Navigation/NAVIGATE',
      routeName,
      params,
    }),
  );
}

function navigateDeep(actions: { routeName: string, params?: NavigationParams }[]) {
  _container.dispatch(
    actions.reduceRight(
      (prevAction, action): any =>
        NavigationActions.navigate({
          type: 'Navigation/NAVIGATE',
          routeName: action.routeName,
          params: action.params,
          action: prevAction,
        }),
      undefined,
    ),
  );
}

function getCurrentRoute(): NavigationRoute | null {
  if (!_container || !_container.state.nav) {
    return null;
  }

  return _container.state.nav.routes[_container.state.nav.index] || null;
}

export default {
  setContainer,
  navigateDeep,
  navigate,
  reset,
  getCurrentRoute,
};
```

and then you can use Navigator service everywhere.

**Comment by keeleycarrigan** [comment]
Maybe I'm misunderstanding something. Is it possible to export the initialized NavigationService to use inside a saga or something? How do you use it everywhere?

**Comment by unexge** [comment]
yes, it just a module. you can import and use everywhere.

```js
import NavigatorService from './services/navigator';

NavigatorService.navigate('Home');
```

**Comment by keeleycarrigan** [comment]
I see. I was thinking about turning this into a class so I could initialize it with multiple navigators if I wanted. I guess I could also have another object to store my instances and allow access to them. Thanks!

**Comment by pietro909** [comment]
Dispatching the action doesn't work for me: it does nothing.
The Navigation is up and running, but when `onClose` is called I can see the log and nonetheless nothing happens :-(

```javascript

const Knipster = StackNavigator(
  {
    Landing: { screen: Landing },
  // ....
   }
)
class Launcher extends React.Component {

  onClose(target) {
    console.warn(`navigate to ${target}`)
    this.navigator.dispatch(
      NavigationActions.navigate({
        type: 'Navigation/NAVIGATE',
        target,
       }),
     )
  }

render() {
  const screenProps = { onClose: this.onClose.bind(this) }

  return (
    <Navigator
      screenProps={screenProps}
      ref={navigatorRef => this.navigator = navigatorRef}
    />
  )
}
```

**Comment by pietro909** [comment]
@unexge I sorted it out: as per [the docs](https://reactnavigation.org/docs/navigators/navigation-actions#Navigate) the action has a different format.

This worked for me:

**Comment by sankalpsingha** [comment]
@unexge s answer is the easiest that I have found so for dispatching to a different navigation even from redux actions. I think something like this should be merged into the main library.

**Comment by keeleycarrigan** [comment]
Yea this solution is great. I made the posted solution into a class so I can save different navigator instances and then store them in another object that allows me to access them anywhere. It's given me pretty good control over navigation without having to pass navigators as props all over the place.

**Comment by digitalmaster** [comment]
@unexge Brilliant!!!!

**Comment by taylorgoolsby** [comment]
@unexge This method does not work if your navigator is mounted with a `navigation` prop. For example:

```
<AppNavigator navigation={addNavigationHelpers({
  dispatch: this.props.dispatch,
  state: this.props.nav,
})} />
```

If anyone does redux integration as described in the [docs](https://reactnavigation.org/docs/guides/redux), then it won't work.

The reason it doesn't work is because the navigator is considered to be not stateful, and `dispatch` will not do anything if it is. Here is the code responsible for this behavior in `createNavigationContainer.js`:

```
dispatch = (action: NavigationAction) => {
  const { state } = this;
  if (!this._isStateful()) {
    return false;
  }
  const nav = Component.router.getStateForAction(action, state.nav);
  if (nav && nav !== state.nav) {
    this.setState({ nav }, () =>
      this._onNavigationStateChange(state.nav, nav, action));
    return true;
  }
  return false;
};
``` 

Is there any reason `dispatch` cannot defer to `this.props.navigation.dispatch`? In other words, is there any reason `dispatch` can't be implemented as is:

```
dispatch = (action: NavigationAction) => {
  const { state } = this;
  if (!this._isStateful()) {
    return this.props.navigation.dispatch(action);
  }
  const nav = Component.router.getStateForAction(action, state.nav);
  if (nav && nav !== state.nav) {
    this.setState({ nav }, () =>
      this._onNavigationStateChange(state.nav, nav, action));
    return true;
  }
  return false;
};
``` 

I've gone ahead and implemented this and did minimal testing, and nothing seems to break. Does anyone know if this is a safe change?

**Comment by dwilt** [comment]
I love the solution that @unexge presented but it seems a little ridiculous that it's this difficult

**Comment by MacKentoch** [comment]
@unexge solution works nicely. Thank you!

But why does [dispatch navigation as docs says](https://reactnavigation.org/docs/navigators/navigation-actions#Navigate) **does not work** to navigate from sub navigators even without redux integration?

The idea of using anywhere in the application (*with tons of navigators and sub navigators*) just something like:
```javascript

import { NavigationActions } from 'react-navigation'

const navigateAction = NavigationActions.navigate({
  routeName: 'Profile',
  params: {},
  action: NavigationActions.navigate({ routeName: 'SubProfileRoute'})
})

this.props.navigation.dispatch(navigateAction)
```
is pretty elegant and concise.

Maybe I'm wrong and did not understand?

**Comment by ThaJay** [comment]
@unexge Would it be wise to make a pull request from your code example.

I agree this is something we should not have to implement ourselves if we only need to navigate 'to' something and 'back'.

The only reason I need this right now is because I have a drawer that is used for more than navigation and it was all working perfectly with react-native-router-flux but then they decide to change everything so I ditched it and implemented react-navigation without redux integration (because that gave me an unreasonable headache.) and this way navigation stays simple like it should be. But then I was unable to open our options page from the drawer.

long story short I have copied this whole navigation module for just one call.

**Comment by adrienjt** [comment]
The NavigatorService solution [above](https://github.com/react-community/react-navigation/issues/1439#issuecomment-303661539) or presented [earlier](https://github.com/react-community/react-navigation/issues/153#issuecomment-277639975) offer nice abstractions around react-navigation, but if you:
- need more flexibility,
- don't want to maintain/extend yet another service,
- and/or are OK with spreading the dependency on react-navigation's NavigationActions,

there's a simpler solution: just export/import the navigator reference:

```js
// App.js

const AppNavigator = StackNavigator(myRouteConfigs);

export let navigatorRef;

export class App extends Component {
  componentDidMount() {
    navigatorRef = this.navigator;
  }

  render() {
    return (
      <AppNavigator ref={nav => { this.navigator = nav; }} />
    );
  }
}
```

```js
// someThunks.js (or anywhereOutsideApp.js)

import {navigatorRef} from 'path/to/App';
import {NavigationActions} from 'react-navigation';

import {someEvent} from './someEvents';
import {someSelector} from './someSelectors';

export const someThunk = () => (dispatch, getState) => {
  dispatch(someEvent()); // this is Redux's dispatch
  const foo = someSelector(getState());
  if (foo) {
    navigatorRef.dispatch(NavigationActions.reset({ // this is react-navigation's dispatch
      index: 0,
      actions: [NavigationActions.navigate({routeName: 'Home'})],
    }));
  }
};
```

**Comment by chefarchitekt** [comment]
@adrienjt Thanks a lot. I manage to use your method successfully. Even I use it to handle Android back button.  Here is my code:

1. I am using Reactotron and desktop version of console debugger
2. I am also using store.dispatch to run certain action method(s) at App's constructor
3. I am using Redux for the whole process EXCEPT for React Navigation with Redux as it is really cumbersome (API is really confusing and its sample and documentation is not helpful)


**App.js**
`import '../reactotron-config';

import React, { Component } from 'react';
import { BackHandler } from 'react-native';
import { Provider } from 'react-redux';
import { createStore, compose, applyMiddleware } from 'redux';
import ReduxThunk from 'redux-thunk';
import { Routes } from './RootNavigator';
import { NavigationActions } from 'react-navigation';
import rootReducer from './data/reducers/rootReducer';
import { checkAuthenticationStatus } from './process/actions/auth/loginActions';


export let navigationRef;
class App extends Component {
    constructor(props) {
        super(props);
        this.store = createStore(
            rootReducer, {}, 
            compose(
                applyMiddleware(ReduxThunk),
                window.devToolsExtension ? window.devToolsExtension() : f => f // if we have this tool we initialize it
            )
        ); 
        this.store.dispatch(checkAuthenticationStatus());
        const initialState = this.store.getState();
        console.log('initialState at App constructor: ', initialState);
        this.handleBackButton = this.handleBackButton.bind(this);
    }

    
    componentDidMount() {
        navigationRef = this.navigator;
        BackHandler.addEventListener('hardwareBackPress', this.handleBackButton);
    }
    
    componentWillUnmount() {
        BackHandler.removeEventListener('hardwareBackPress', this.handleBackButton);
    }

    handleBackButton() {
        navigationRef.dispatch(NavigationActions.navigate({ 
            routeName: 'Main'
        }));
    }

    render() {
        return (
            <Provider store={this.store}>
                <Routes ref={nav => { this.navigator = nav; }} />
            </Provider>
        );
    }
}

export default App;

**`loginAction.js**
`
import { NavigationActions } from 'react-navigation';
import { navigationRef } from '../../../App';

import types....

const accountLoginAsync = (dispatch, encodedLoginData, formLoginData) => {
    console.log('action: encodedLoginData');
    console.log(encodedLoginData);
    fetch('SOME_URL', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: encodedLoginData
      })
      .then((response) => response.json())
      .then((responseJson) => {
            if (responseJson.Status === 'Error') {
                loginUserFailed(dispatch, responseJson.Record);
                httpErrorDetail(dispatch, responseJson.Record); // response.Record = 'Api Error'
            } else {
                loginUserSuccess(dispatch, formLoginData, responseJson);
                setCurrentUser(dispatch, formLoginData, responseJson.Record);
            }
        })
      .catch((error) => {
            httpErrorDetail(dispatch, error.response);
            loginUserFailed(dispatch, error.response.data);
        });
  };

  const loginUserSuccess = (dispatch, formLoginData, responseData) => {
        dispatch({
            type: LOGIN_USER_SUCCESS,
            payload: { formLoginData, responseData }
        });
        navigationRef.dispatch(NavigationActions.navigate({ // this is react-navigation's dispatch
            routeName: 'Main'
        }));
 };
 
 const loginUserFailed = (dispatch, errorData) => {
     dispatch({
         type: SERVER_LOGIC_ERRORS,
         payload: errorData
     });
 };
 
 const httpErrorDetail = (dispatch, errorDetail) => {
     dispatch({
         type: HTTP_ERRORS,
         payload: errorDetail
     });
 };


....others

**Comment by adrienjt** [comment]
@chefarchitekt why do you handle the back button? react-navigation should handle it for you if it is not integrated with Redux.

**Comment by msageryd** [comment]
@unexge You are my hero. Thank you so much for your solution. It saved my day!

**Comment by mtx62** [comment]
@unexge I really don't know what you did, but you are a HERO!

**Comment by Vamshikrishna209** [comment]
@unexge 
Hi , I am using react-navigation for routing between the screens/components 
I have written the fetch api function outside the component keeping it available to all the components
So i need to check if the user is logged in when the api call response is available , 
So if the user token has been expired , I need to route him to login page from externaljs file , Can you please help me how it is done from external js file (Not a component).

**Comment by unexge** [comment]
@Vamshikrishna209  here is an example:

```js
import NavigatorService from './services/navigator';

// ...

if (userTokenIsExpired()) {
    NavigatorService.reset('Login');
}
```

**Comment by dorshay6** [comment]
Another soultion in here:
```
import React from 'react';
let navigatorRef = null;

export const screenHOC = (WrappedComponent) => {
    const Component = (props) => {
        navigatorRef = props.navigator;
        return (<WrappedComponent {...props} />);
    }
    Component.navigatorStyle = WrappedComponent.navigatorStyle;

    return Component;
}

export const pushScreen = (payload) => navigatorRef.push(payload);
```


```
class App extends Component {
  static navigatorStyle = {
    navBarBackgroundColor: '#2a262c',
    navBarTextColor: '#fff',
    screenBackgroundColor: '#14091a',
  };
 render () {
 }
}
export default screenHOC(App);
```

**Comment by brentvatne** [comment]
this is a docs issue, moved it to https://github.com/react-navigation/react-navigation.github.io/issues/29 -- thanks everyone for providing answers!

**Comment by digitalmaster** [comment]
hmm.. Yea, i guess we could just document one of the work arounds we've collected here but I think ideally this should be a feature - not requiring a work around - and we would then of course document that. 

That said, if a proper implementation is non-trivial then this could certainly be the documented way to accomplish this.

**Comment by i6mi6** [comment]
Using `react-navigation@1.0.0-beta.11`
```

import { addNavigationHelpers } from 'react-navigation';

class Root extends Component {
  render() {
    const nav = addNavigationHelpers({
      dispatch: this.props.dispatch,
      state: this.props.nav,
    })
    return (
      <StackNavigation navigation={nav}/>
    )
  }
}

const mapStateToProps = state => {
  return { nav: state.nav }
};

export default connect(mapStateToProps)(Root);
```

then call it in Sagas 
`yield put(NavigationActions.navigate(...))`
or Thunk
`dispatch(NavigationActions.navigate(...))`

**Comment by brentvatne** [comment]
@digitalmaster - makes sense, feel free to open a RFC to describe how the feature should work: https://github.com/react-navigation/rfcs

I personally would not use this feature so someone else will need to draft the RFC

**Comment by nthgol** [comment]
If you are looking to dispatch navigation actions inside a redux action creator, but don't want the  headache of the current redux integration API, another workaround, though not great, is to pass the navigation action as a callback:
`this.props.myActionCreator(() => this.props.navigation.goBack());` inside your component
`export const myActionCreator = (callback) => {
     callback();
}` inside your action creator

**Comment by pedrogoes** [comment]
I made this solution: https://www.npmjs.com/package/react-native-navigation-actions

**Comment by djmalice** [comment]
@pedrogoes : Thanks. Makes life easy for newbies like me. Works like a charm!!

**Comment by Murrayee** [comment]
@taylorgoolsby You are my hero too . Thank you so much for your solution. It saved my day!

**Comment by xlcrr** [comment]
Sorry but what is
```
import NavigatorService from './services/navigator';
```
?

**Comment by ukaric** [comment]
I can confirm that without integrating navigation state into Redux that @pedrogoes package works like charm. I'm dispatching nav actions from action creators with no issue. Great job it is by far easiest solution.

**Comment by giacomocerquone** [comment]
@unexge That service can be considered to emulate the behaviour of navigating through props in react-navigation? Or it is computationally more expensive?
What I mean is that everytime I need to navigate deep in my nested navigators structure, with this service, I'd spawn a navigation action that starts from the root and bubbles down until the bottom one.

Is this good?

**Comment by unexge** [comment]
@giacomocerquone i don't know about last versions of react-navigation but there should be no overhead.

**Comment by giacomocerquone** [comment]
@unexge thanks a lot. Benchmarked naively this with a console log in the component rendered and yes, the same components are rendered using these two ways.

I changed your code though in this way:

```
function navigateDeep(actions: { routeName: string, params?: NavigationParams }[]) {
  _navigator.dispatch(actions.reduce(
    (prevAction, action): any =>
      NavigationActions.navigate({
        routeName: action.routeName || action,
        params: action.params,
        action: prevAction,
      }),
    undefined,
  ));
}
```

In this way I'm able to navigate like this:
`navService.navigateDeep(['Route1', 'Route2']);`
And if I need param I can always throw an object in there.

Ps: can't understand why you preferred a reduceright instead of the plain reduce... I find it obvious that's better to express routes where I want to navigate to from left to right

**Comment by uinz** [comment]
hey guys.

I using `"react-navigation": "^2.6.2"`

how can I push (not navigate) same routerName without `props.navigation`?

**Comment by ayushkamadji** [comment]
Hi fellas, I'm obviously missing something here. How does `_container` get it's reference navigator outside of App.js (or index.js or wherever `setContainer` was called)?? Have I misunderstood js modules scope all this time??

**Comment by ayushkamadji** [comment]
I was expecting modules to have their own scope, this is common in other languages. I have been using js for well over a year now without ever encountering this particular aspect. For those who have the same confusion I suggest googling for "js modules are singleton". Sorry for rubber ducking on an issue.

**Comment by giacomocerquone** [comment]
modules do have their own scope (infact they can be considered as singletons)... I believe you're really off the road. Not having a scope means that it falls down to the global scope, but that's not the case @ayushkamadji

**Comment by NotTooReact89** [comment]
@unexge did you figure out the flow type for the navigatorRef and container? Using Object is not recommended flow type.

**Comment by forrest23** [comment]
This worked for me, you get access to navigator through refs and you can use it then anywhere in the app to dispatch https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html

**Comment by SaeedZhiany** [comment]
I'm using typescript in my project and I'm stuck with the type of the ref, does anyone know what type I should set for the ref Object in this code? ([code sample is taken from docs](https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html))

```
// NavigationService.js

import { NavigationActions } from 'react-navigation';

let _navigator;  // <- what is the exact type of this ref?

function setTopLevelNavigator(navigatorRef) {
  _navigator = navigatorRef;
}
```

**Comment by dwilt** [comment]
with V3 I'm doing this:

**navigation.service.ts**
```ts
export const navigator = createRef<NavigationContainerComponent>();
```

and then in my routes..
```ts
<RootStack ref={NavigationService.navigator} />
```

**Comment by SaeedZhiany** [comment]
@dwilt 
Is possible to set the ref using the `setTopLevelNavigator` function, I prefer to not exporting the `navigator ` to the outside of the file.

**Comment by SaeedZhiany** [comment]
I found what I want:

```
let instanceRef: NavigationContainerComponent;

export function setNavigatorRef(instance: NavigationContainerComponent) {
  instanceRef = instance;
}
```
and using it like below:

```
<AppContainer ref={setNavigatorRef} />
```

**Comment by sregg** [comment]
I'm using this `NavigationService` so I can dispatch navigation actions outside of component code (e.g. redux actions, etc...).

But one of my colleague started to use it in components too, instead of using `this.props.navigation`.
Is there any benefits from using `this.props.navigation` vs `NavigationService`?

**Comment by reactdevrocks** [comment]
hi @sregg please tell me the example with redux as i am not able to navigate without props.

export default class App extends Component
{
  render() {
      return (
          <Provider store={store} >
            <AppNavigator
              ref={navigatorRef => {
                NavigationService.setContainer(navigatorRef);
              }}
            />
          </Provider>
      )
  }
}

**Comment by btoo** [comment]
is there a type safe way we can have top-level `navigate`?

**Comment by esdotzed** [comment]
> Sorry but what is
> 
> ```
> import NavigatorService from './services/navigator';
> ```
> 
> ?

https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html

**Comment by nehresma** [comment]
@uinz 
> how can I push (not navigate) same routerName without `props.navigation`?

I know it's been a while since you asked about this but here's a way you can do it (with react navigation < 5).  Essentially you need to set the type on the action to `Navigation/PUSH`:

```typescript
    let action = NavigationActions.navigate({
      routeName: '.....',
      params: {.....},
    }) as NavigationAction;
    action.type = 'Navigation/PUSH';
    myNavigatorRef.dispatch(action);
```

**Comment by pranav-pandey** [comment]
> ```js
> routeName
> ```

Thanks for the solution. It works by the ref, but it always fails in the saga with 
```Error: You can't put (a.k.a. dispatch from saga) frozen actions.
We have to define a special non-enumerable property on those actions for scheduling purposes.
Otherwise you wouldn't be able to communicate properly between sagas & other subscribers (action ordering would become far less predictable).
If you are using redux and you care about this behaviour (frozen actions),
then you might want to switch to freezing actions in a middleware rather than in action creator.
Example implementation:
const freezeActions = store => next => action => next(Object.freeze(action))
    at check (io-1d6eccda.js:37)
    at io-1d6eccda.js:132
    at redux-saga-core.dev.cjs.js:476
    at exec (redux-saga-core.dev.cjs.js:35)
    at flush (redux-saga-core.dev.cjs.js:91)
    at asap (redux-saga-core.dev.cjs.js:50)
    at runPutEffect (redux-saga-core.dev.cjs.js:472)
    at runEffect (redux-saga-core.dev.cjs.js:1208)
    at digestEffect (redux-saga-core.dev.cjs.js:1275)
    at next (redux-saga-core.dev.cjs.js:1165)```

Any insights?

**Comment by pranav-pandey** [comment]
https://github.com/react-navigation/react-navigation/issues/1439#issuecomment-340293063

This works like a charm, but on usage in saga, throws me a freeze actions error
```Error: You can't put (a.k.a. dispatch from saga) frozen actions.
We have to define a special non-enumerable property on those actions for scheduling purposes.
Otherwise you wouldn't be able to communicate properly between sagas & other subscribers (action ordering would become far less predictable).
If you are using redux and you care about this behaviour (frozen actions),
then you might want to switch to freezing actions in a middleware rather than in action creator.
Example implementation:
const freezeActions = store => next => action => next(Object.freeze(action))
    at check (io-1d6eccda.js:37)
    at io-1d6eccda.js:132
    at redux-saga-core.dev.cjs.js:476
    at exec (redux-saga-core.dev.cjs.js:35)
    at flush (redux-saga-core.dev.cjs.js:91)
    at asap (redux-saga-core.dev.cjs.js:50)
    at runPutEffect (redux-saga-core.dev.cjs.js:472)
    at runEffect (redux-saga-core.dev.cjs.js:1208)
    at digestEffect (redux-saga-core.dev.cjs.js:1275)
    at next (redux-saga-core.dev.cjs.js:1165)```

**Comment by NachosSs** [comment]
Hi, how to get there with react navigation 5 ? We can't set a ref on Stack.Navigator (Function components cannot be given refs)

**Comment by giacomocerquone** [comment]
@NachosSs https://reactnavigation.org/docs/en/next/navigating-without-navigation-prop.html

Read the documentation :)

**Comment by NachosSs** [comment]
Sorry, I miss this item. It's works, Thanks

**Comment by minhnguyenvan95** [comment]
> you can pass your top-level navigator ref to a service, and dispatch actions from that service.
> 
> here is mine:
> 
> ```js
> // App.js
> 
> import NavigatorService from './services/navigator';
> 
> const Navigator = StackNavigator({ /* ... */ })
> 
> class App extends Component {
>   // ...
> 
>   render(): {
>     return (
>       <Navigator
>         ref={navigatorRef => {
>           NavigatorService.setContainer(navigatorRef);
>         }}
>       />
>     );
>   }
> }
> ```
> 
> ```js
> // services/navigator.js
> // @flow
> 
> import { NavigationActions } from 'react-navigation';
> import type { NavigationParams, NavigationRoute } from 'react-navigation';
> 
> let _container; // eslint-disable-line
> 
> function setContainer(container: Object) {
>   _container = container;
> }
> 
> function reset(routeName: string, params?: NavigationParams) {
>   _container.dispatch(
>     NavigationActions.reset({
>       index: 0,
>       actions: [
>         NavigationActions.navigate({
>           type: 'Navigation/NAVIGATE',
>           routeName,
>           params,
>         }),
>       ],
>     }),
>   );
> }
> 
> function navigate(routeName: string, params?: NavigationParams) {
>   _container.dispatch(
>     NavigationActions.navigate({
>       type: 'Navigation/NAVIGATE',
>       routeName,
>       params,
>     }),
>   );
> }
> 
> function navigateDeep(actions: { routeName: string, params?: NavigationParams }[]) {
>   _container.dispatch(
>     actions.reduceRight(
>       (prevAction, action): any =>
>         NavigationActions.navigate({
>           type: 'Navigation/NAVIGATE',
>           routeName: action.routeName,
>           params: action.params,
>           action: prevAction,
>         }),
>       undefined,
>     ),
>   );
> }
> 
> function getCurrentRoute(): NavigationRoute | null {
>   if (!_container || !_container.state.nav) {
>     return null;
>   }
> 
>   return _container.state.nav.routes[_container.state.nav.index] || null;
> }
> 
> export default {
>   setContainer,
>   navigateDeep,
>   navigate,
>   reset,
>   getCurrentRoute,
> };
> ```
> 
> and then you can use Navigator service everywhere.

for navigation 5

```
import { CommonActions } from '@react-navigation/core';

let _container; // eslint-disable-line

function setContainer(container) {
  _container = container;
}

function reset(routeName, params) {
  _container.dispatch(
    CommonActions.reset({
      index: 0,
      actions: [
        CommonActions.navigate({
          name: routeName,
          params: params,
        }),
      ],
    }),
  );
}

function navigate(routeName, params) {
  _container.dispatch(
    CommonActions.navigate({
      name: routeName,
      params: params,
    })
  );
}

function navigateDeep(actions) {
  _container.dispatch(
    actions.reduceRight(
      (prevAction, action) =>
        CommonActions.navigate({
          name: routeName,
          params: params,
        }),
      undefined,
    ),
  );
}

function getCurrentRoute() {
  if (!_container || !_container.state.nav) {
    return null;
  }

  return _container.state.nav.routes[_container.state.nav.index] || null;
}

export default {
  setContainer,
  navigateDeep,
  navigate,
  reset,
  getCurrentRoute,
};
```

**Comment by rob5408** [comment]
For @minhnguyenvan95 's RN 5 solution I had to import `CommonActions` from `@react-navigation/native` in order to work.
